### PR TITLE
Update hashbrown to 0.15

### PR DESCRIPTION
**Pull request by clarfonthey**
Updating dependencies; adopted version of #15696. (Supercedes #15696.)

Long answer: hashbrown is no longer using ahash by default, meaning that we can't use the default-hasher methods with ahasher. So, we have to use the longer-winded versions instead. This takes the opportunity to also switch our default hasher as well, but without actually enabling the default-hasher feature for hashbrown, meaning that we'll be able to change our hasher more easily at the cost of all of these method calls being obnoxious forever.

One large change from 0.15 is that `insert_unique_unchecked` is now `unsafe`, and for cases where unsafe code was denied at the crate level, I replaced it with `insert`.

## Migration Guide

`bevy_utils` has updated its version of `hashbrown` to 0.15 and now defaults to `foldhash` instead of `ahash`. This means that if you've hard-coded your hasher to `bevy_utils::AHasher` or separately used the `ahash` crate in your code, you may need to switch to `foldhash` to ensure that everything works like it does in Bevy.

### CODE REVIEW

**Comment by mnmaita** [comment]
@clarfonthey looks like there's a formatting issue somewhere, could you try running the formatter and push again?

**Comment by clarfonthey** [comment]
Yeah, I was running into some difficulties fixing the lints but will try and fix this today or tomorrow.

**Comment by clarfonthey** [comment]
Rereading the changelog, I misunderstood what hashbrown did and they actually switched to a better hash that we should probably use instead. Gonna probably find a way to factor that in.

File: crates/bevy_ecs/src/bundle.rs
Comment (clarfonthey) [inline crates/bevy_ecs/src/bundle.rs:1425]: I'm adding this comment to appease the lint, but I actually don't know if we do know if the ID is unique.

File: crates/bevy_reflect/Cargo.toml
Comment (clarfonthey) [inline crates/bevy_reflect/Cargo.toml:39]: This felt like the best approach: rather than re-exporting foldhash in `bevy_utils` and further bloating up that crate, just re-import it here and add impls for it.

File: crates/bevy_utils/Cargo.toml
Comment (clarfonthey) [inline crates/bevy_utils/Cargo.toml:24]: This is to explicitly exclude `default-hasher` here. It gives us the flexibility to change our hasher whenever since nobody will be relying on the default hasher, even though it *is* a pain to do this.

**Comment by clarfonthey** [comment]
This should be ready for review, modulo whatever tiny nitpicks CI throws at me.

Not sure if the migration guide is adequate: not sure how much we care about bevy_utils' public API, as far as like, whether people should be using it or not.

File: crates/bevy_reflect/Cargo.toml
Comment (bushrat011899) [reply crates/bevy_reflect/Cargo.toml:39]: My only critique here would be to make the `foldhash` dependency optional, keeping inline with how other reflection implementations for 3rd party types are done. Not a blocker, just a comment.

**Review by bushrat011899 — APPROVED** [review]

**Comment by hymm** [comment]
What is better than foldhash vs hash? Is it really worth taking the extra dependency?

**Comment by clarfonthey** [comment]
> What is better about foldhash vs ahash? Is it really worth taking the extra dependency?

Note: it's not an extra dependency, just a different dependency. Since hashbrown is using it now, we'd actually be taking an extra dependency by continuing to use ahash.

But supposedly foldhash performs better than ahash, which seems pretty worth it.

File: crates/bevy_reflect/Cargo.toml
Comment (clarfonthey) [reply crates/bevy_reflect/Cargo.toml:39]: I didn't make it optional only because it isn't actually optional: `bevy_utils` always includes it, so, putting it optional here would just give people the option to accidentally remove impls for it when they need them.

File: crates/bevy_utils/Cargo.toml
Comment (mnmaita) [reply crates/bevy_utils/Cargo.toml:24]: FYI `0.15.0` was yanked so you either want to use `0.15.1` (or `0.15` maybe?).

**Comment by clarfonthey** [comment]
I figure this will be a post-0.15 change, and so will likely have more merge conflicts going forward, but I'm going to try and at least get the current conflicts resolved. Seems to be taking a few tries.

File: crates/bevy_utils/Cargo.toml
Comment (clarfonthey) [reply crates/bevy_utils/Cargo.toml]: I actually marked it as 0.15.1 here because 0.15.0 was un-yanked, but may have some advisories filed for it, even though none of them affect us. Don't want to accidentally trigger any lints.

**Comment by clarfonthey** [comment]
(I fixed everything on my side, but an advisory issue is showing up that seems unrelated to this PR.)

**Comment by BenjaminBrienen** [comment]
Fixes https://github.com/bevyengine/bevy/issues/16478

**Review by BenjaminBrienen — APPROVED** [review]

**Comment by clarfonthey** [comment]
Rebasing to account for the new graph types in ECS: I think that once someone looks at that version, we should mark this as ready for final review, since I'd like to get this out before any more big PRs if possible.

File: crates/bevy_ecs/src/schedule/graph/graph_map.rs
Comment (clarfonthey) [inline crates/bevy_ecs/src/schedule/graph/graph_map.rs:403]: Here's some cursed Rust knowledge:

Yes, `<DiGraph>` and `DiGraph` mean different things here. `<DiGraph>` is equivalent to `<DiGraph<FixedHasher>>` (using the defaults) whereas `DiGraph` is equivalent to `<DiGraph<_>>` (inferring the parameter.

File: crates/bevy_ecs/src/schedule/graph/graph_map.rs
Comment (clarfonthey) [inline crates/bevy_ecs/src/schedule/graph/graph_map.rs:23]: I'm not sure if the choice to use `BuildHasherDefault<AHasher>` here was heavily motivated, but it felt prudent to use the same default as everywhere else, in particular, using the chosen fixed hasher.

**Comment by clarfonthey** [comment]
(Apologies for re-requesting reviews from people: I figure that it's better to "dismiss" the old approvals whenever I make changes, but there's no expectation to review if you're busy.)

File: crates/bevy_ecs/src/schedule/graph/graph_map.rs
Comment (bushrat011899) [reply crates/bevy_ecs/src/schedule/graph/graph_map.rs:23]: Can confirm I just chose a hasher without particular concern. Consistency is a good reason to change it!

File: crates/bevy_ecs/src/schedule/graph/graph_map.rs
Comment (bushrat011899) [reply crates/bevy_ecs/src/schedule/graph/graph_map.rs:403]: Interesting! Is that a deliberate feature of the language or just a quirk in the compiler?

**Review by bushrat011899 — APPROVED** [review]
Nice work! As a heads-up, I believe the `no_std` of `bevy_reflect` #16256 will merge soon too, which might add some new conflicts (I don't know for certain, might merge well!). If there are conflicts please feel free to @ me or re-request my review!

**Comment by clarfonthey** [comment]
Since the PR @bushrat011899 mentioned has been merged now, I think it's probably okay to push this through soon. Will wait for CI before I bother people, though.

(Yes, I know that this comment counts as bothering people. Shh)

File: crates/bevy_ecs/src/bundle.rs
Comment (mnmaita) [reply crates/bevy_ecs/src/bundle.rs:1425]: I´ve checked the places where `Bundles::init_dynamic_info` was being used and most of them are internal facing, but I found `EntityWorldMut::insert_by_ids`, `EntityWorldMut::retain` and `EntityWorldMut::clear` which might be public facing I think?

`HashMap::insert_unique_unchecked` says:

```rust
    /// This operation is safe if a key does not exist in the map.
    ///
    /// However, if a key exists in the map already, the behavior is unspecified:
    /// this operation may panic, loop forever, or any following operation with the map
    /// may panic, loop forever or return arbitrary result.
```

So I'm unsure if we should say something like the caller ensures the uniqueness, and even add some notes in the `EntityWorldMut` methods if applicable? I guess I'm thinking out loud. 😅

**Review by mnmaita — APPROVED** [review]
This looks good codewise, I'm just leaving a small comment.

File: crates/bevy_ecs/src/bundle.rs
Comment (clarfonthey) [reply crates/bevy_ecs/src/bundle.rs:1425]: So, I'm honestly fine just changing this to `insert` and leaving it at that :p

**Review by bushrat011899 — APPROVED** [review]
Nice work managing this change!

File: crates/bevy_ecs/src/bundle.rs
Comment (clarfonthey) [reply crates/bevy_ecs/src/bundle.rs]: Okay, so, I double-checked and the bundle ID *is* always increasing here, and without concurrency, we don't need to worry about data races. So, I updated the comment to clarify it is actually unique and leaving it.

**Comment by clarfonthey** [comment]
Final changes made:

* Update "ID is unique" comment to actually clarify why ID is unique
* Normalise `let x: HashMap<_, _> = HashMap::` to `let x = <HashMap<_, _>>::` and similarly for `HashSet`, to make things less verbose

**Comment by clarfonthey** [comment]
(If someone could mark this as ready for final review, I'd appreciate it. I think this should qualify as ready.)

**Comment by clarfonthey** [comment]
(I'll try and take a look at the failures soon)

**Review by BenjaminBrienen — APPROVED** [review]

**Comment by clarfonthey** [comment]
Ah, fun: the singular PR that got in the queue before this one didn't result in any merge conflicts, but it did add more uses of `insert_unique_unchecked` that now need unsafe blocks :(

**Comment by clarfonthey** [comment]
(kind of shocked that `ci compile` didn't catch that, tbh)

**Comment by clarfonthey** [comment]
(Okay, this time I waited until CI passed. Should be ready to merge now. <3)

**Comment by alice-i-cecile** [comment]
Thanks for shepherding this across the finish line <3

**Comment by clarfonthey** [comment]
So, I had no idea about these patches that were being run in CI, but they seem pretty brittle-- it would be much better (IMHO) to use `#[cfg]` or `cfg!` to enable/disable these features than use patches, IMHO, so that at least they would be visible to people searching through the code. That said, I'll try and see what I can do.

**Comment by clarfonthey** [comment]
I also have *genuinely* no idea why that patch failed on this PR. It appears to be referencing an earlier change where the patch should have broken, but it somehow did not trigger until now. And there were no recent changes to the CI job, either. It just makes absolutely no sense why this PR would be the one to trigger it, but at least, it did, and I fixed it.

**Pull request merged**
